{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from prophet import Prophet\n",
    "from sklearn.metrics import mean_squared_error, mean_absolute_error\n",
    "import pickle\n",
    "import os\n",
    "\n",
    "\n",
    "# # Использование класса\n",
    "# sales_data = ...  # Загрузите данные о продажах\n",
    "# calendar_data = ...  # Загрузите данные календаря\n",
    "# prices_data = ...  # Загрузите данные о ценах\n",
    "\n",
    "# preprocessor = DataPreprocessor(sales_data, calendar_data, prices_data)\n",
    "\n",
    "class RetailSalesPredictor:\n",
    "    def __init__(self, sales_data, calendar_data, prices_data):\n",
    "        self.sales_data = sales_data\n",
    "        self.calendar_data = calendar_data\n",
    "        self.prices_data = prices_data\n",
    "                                                                                # Мы заранее выбрали STORE 1\n",
    "    def preprocess_data(self, item_id, sales_data, calendar_data, prices_data): # Выбираем товар (item_id) из магазина STORE 1, подаем  df: sales, calendar, price\n",
    "        item = str(item_id)\n",
    "        sales = sales_data\n",
    "\n",
    "        ### Обрабатываем основную таблицу sales_data и отбираем ряд для конкретного товара\n",
    "\n",
    "        customer = sales[(sales['store_id'] == 'STORE_1') & (sales['item_id'] == f'STORE_1_{item}')]\n",
    "        customer = pd.merge(customer, calendar_data, on='date_id', how='inner')\n",
    "        customer = customer.rename(columns={'date': 'ds', 'cnt': 'y'})\n",
    "        customer = customer[['ds', 'y']]\n",
    "\n",
    "        customer['ds'] = pd.to_datetime(customer['ds'])\n",
    "\n",
    "        ### Обрабатываем праздники\n",
    "\n",
    "        # Извлекаем даты и названия праздников\n",
    "        holidays_1 = calendar_data[['date', 'event_name_1']].dropna().rename(columns={'date': 'ds', 'event_name_1': 'holiday'})\n",
    "        holidays_2 = calendar_data[['date', 'event_name_2']].dropna().rename(columns={'date': 'ds', 'event_name_2': 'holiday'})\n",
    "\n",
    "        # Объединяем оба набора данных\n",
    "        holidays = pd.concat([holidays_1, holidays_2])\n",
    "\n",
    "        # Убедимся, что столбец ds имеет тип datetime\n",
    "        holidays['ds'] = pd.to_datetime(holidays['ds'])\n",
    "\n",
    "        # Создание DataFrame с будущими датами праздников\n",
    "        future_years = [2016, 2017]\n",
    "        future_holidays = []\n",
    "\n",
    "        for year in future_years:\n",
    "            for holiday in holidays['holiday'].unique():\n",
    "                future_holidays.append({\n",
    "                    'holiday': holiday,\n",
    "                    'ds': pd.to_datetime(f\"{year}-01-01\"),  # Здесь нужно указать соответствующие даты праздников\n",
    "                    'lower_window': 0,\n",
    "                    'upper_window': 1\n",
    "                })\n",
    "\n",
    "        future_holidays_df = pd.DataFrame(future_holidays)\n",
    "        # Объединяем текущие и будущие праздники\n",
    "        all_holidays = pd.concat([holidays, future_holidays_df])\n",
    "\n",
    "        ###  Обрабатываем цены\n",
    "\n",
    "        item_prices = prices_data[prices_data['item_id'] == f'STORE_1_{item}'][['wm_yr_wk', 'sell_price']]\n",
    "\n",
    "        # Убедимся, что wm_yr_wk в формате целого числа\n",
    "        item_prices['wm_yr_wk'] = item_prices['wm_yr_wk'].astype(int)\n",
    "\n",
    "        # Переименовываем столбцы для удобства\n",
    "        item_prices = item_prices.rename(columns={'wm_yr_wk': 'week', 'sell_price': 'price'})\n",
    "\n",
    "        # Добавляем даты в данные о ценах, если у нас есть календарь с недельным номером и датами\n",
    "        calendar_data['wm_yr_wk'] = calendar_data['wm_yr_wk'].astype(int)\n",
    "        item_prices = item_prices.merge(calendar_data[['wm_yr_wk', 'date']], left_on='week', right_on='wm_yr_wk', how='left')\n",
    "        item_prices = item_prices.rename(columns={'date': 'ds'}).drop(columns=['wm_yr_wk'])\n",
    "\n",
    "        # Убедимся, что столбец ds имеет тип datetime\n",
    "        item_prices['ds'] = pd.to_datetime(item_prices['ds'])\n",
    "\n",
    "        # Убедимся, что столбец ds имеет тип datetime и в customer\n",
    "        customer['ds'] = pd.to_datetime(customer['ds'])\n",
    "\n",
    "        # Объединение данных о продажах и ценах\n",
    "        customer = customer.merge(item_prices[['ds', 'price']], on='ds', how='left')\n",
    "\n",
    "        return f'STORE_1_{item}', customer, all_holidays\n",
    "    \n",
    "    def train_model(self, df, holydays, test_size = 30):\n",
    "        train_data = df.iloc[:test_size]\n",
    "        test_data = df.iloc[test_size:]\n",
    "\n",
    "        # Создание и обучение модели Prophet с праздниками и регрессором\n",
    "        model = Prophet(holidays=holydays)\n",
    "        model.add_regressor('price')\n",
    "\n",
    "        # Обучение модели\n",
    "        model.fit(train_data)\n",
    "        return model, train_data, test_data\n",
    "\n",
    "    def predict_and_evaluate(self, model, customer, period, test_data ):\n",
    "        # Создание прогноза для тестовой выборки\n",
    "        future = model.make_future_dataframe(periods=period)\n",
    "        # Объединяем данные о будущем с будущими ценами (предполагается, что у вас есть прогноз цен или просто используем последние известные цены)\n",
    "        future = future.merge(customer[['ds', 'price']], on='ds', how='left')\n",
    "\n",
    "        # Предсказание\n",
    "        forecast = model.predict(future)\n",
    "\n",
    "        # Выборка последних 30 дней прогноза\n",
    "        predicted = forecast[['ds', 'yhat']].tail(period)\n",
    "\n",
    "        # Метрики точности прогноза\n",
    "        actual = test_data['y'].values\n",
    "        predicted_values = predicted['yhat'].values\n",
    "\n",
    "        mse = mean_squared_error(actual, predicted_values)\n",
    "        mae = mean_absolute_error(actual, predicted_values)\n",
    "\n",
    "        print(f'Mean Squared Error: {mse}')\n",
    "        print(f'Mean Absolute Error: {mae}')\n",
    "\n",
    "        # Визуализация прогноза\n",
    "        fig1 = model.plot(forecast)\n",
    "        plt.show()\n",
    "\n",
    "        # Визуализация компонентов прогноза\n",
    "        fig2 = model.plot_components(forecast)\n",
    "        plt.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "sales_data = pd.read_csv('shop_sales.csv')  # Загрузите данные о продажах\n",
    "calendar_data = pd.read_csv('shop_sales_dates.csv')   # Загрузите данные календаря\n",
    "prices_data = pd.read_csv('shop_sales_dates.csv')   # Загрузите данные о ценах"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_class = RetailSalesPredictor(sales_data, calendar_data, prices_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'item_id'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m~/anaconda3/envs/timeseries/lib/python3.9/site-packages/pandas/core/indexes/base.py:3805\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3804\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 3805\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcasted_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3806\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "File \u001b[0;32mindex.pyx:167\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mindex.pyx:196\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:7081\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:7089\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'item_id'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[9], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m store_item, customer_table, all_holidays_table \u001b[38;5;241m=\u001b[39m \u001b[43mdata_class\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mpreprocess_data\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m252\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msales_data\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mcalendar_data\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mprices_data\u001b[49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn[6], line 67\u001b[0m, in \u001b[0;36mRetailSalesPredictor.preprocess_data\u001b[0;34m(self, item_id, sales_data, calendar_data, prices_data)\u001b[0m\n\u001b[1;32m     63\u001b[0m all_holidays \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mconcat([holidays, future_holidays_df])\n\u001b[1;32m     65\u001b[0m \u001b[38;5;66;03m###  Обрабатываем цены\u001b[39;00m\n\u001b[0;32m---> 67\u001b[0m item_prices \u001b[38;5;241m=\u001b[39m prices_data[\u001b[43mprices_data\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mitem_id\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m \u001b[38;5;241m==\u001b[39m \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mSTORE_1_\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mitem\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m'\u001b[39m][[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mwm_yr_wk\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124msell_price\u001b[39m\u001b[38;5;124m'\u001b[39m]]\n\u001b[1;32m     69\u001b[0m \u001b[38;5;66;03m# Убедимся, что wm_yr_wk в формате целого числа\u001b[39;00m\n\u001b[1;32m     70\u001b[0m item_prices[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mwm_yr_wk\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m item_prices[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mwm_yr_wk\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39mastype(\u001b[38;5;28mint\u001b[39m)\n",
      "File \u001b[0;32m~/anaconda3/envs/timeseries/lib/python3.9/site-packages/pandas/core/frame.py:4102\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   4100\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcolumns\u001b[38;5;241m.\u001b[39mnlevels \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m1\u001b[39m:\n\u001b[1;32m   4101\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_getitem_multilevel(key)\n\u001b[0;32m-> 4102\u001b[0m indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   4103\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_integer(indexer):\n\u001b[1;32m   4104\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m [indexer]\n",
      "File \u001b[0;32m~/anaconda3/envs/timeseries/lib/python3.9/site-packages/pandas/core/indexes/base.py:3812\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3807\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(casted_key, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;129;01mor\u001b[39;00m (\n\u001b[1;32m   3808\u001b[0m         \u001b[38;5;28misinstance\u001b[39m(casted_key, abc\u001b[38;5;241m.\u001b[39mIterable)\n\u001b[1;32m   3809\u001b[0m         \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28many\u001b[39m(\u001b[38;5;28misinstance\u001b[39m(x, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m casted_key)\n\u001b[1;32m   3810\u001b[0m     ):\n\u001b[1;32m   3811\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m InvalidIndexError(key)\n\u001b[0;32m-> 3812\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01merr\u001b[39;00m\n\u001b[1;32m   3813\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[1;32m   3814\u001b[0m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[1;32m   3815\u001b[0m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[1;32m   3816\u001b[0m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[1;32m   3817\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'item_id'"
     ]
    }
   ],
   "source": [
    "store_item, customer_table, all_holidays_table = data_class.preprocess_data(252, sales_data, calendar_data, prices_data)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "timeseries",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
